File: ueb4/scala-server/src/main/scala/Connection.scala
import akka.actor._

object Connection {
  case class Start(client: ActorRef)
  case object Stop
}

class Connection extends Actor with ActorLogging {
  def receive = {
    case Connection.Start(client) =>
      log.info("new client connected")
      context.become(connected(client))
  }

  def connected(client: ActorRef): Receive = {
    case TimestampedMessage(t,Ping) =>
      println(t,Ping)
      client ! TimestampedMessage(Pong)
    case Connection.Stop =>
      log.info("connection closed")
      context.stop(self)
  }
}

File: ueb4/haskell-server/Connection.hs
{-# LANGUAGE LambdaCase #-}
module Connection where

import Prelude hiding (log)
import Hakka.Actor
import Messages
import Control.Concurrent.MVar
import Control.Concurrent
import Control.Exception

connection :: Int -> ActorRef ServerMessage -> ActorIO ServerMessage (ActorRef ServerMessage)
connection n univ = actor ("connection" ++ show n) receive where
  receive = \case
    Connect out res -> do
      log Info "new client connected"            
      self >>= liftIO . putMVar res
      become $ connected out
    Disconnect _ -> do
      univ ! Disconnect n
      log Info "connection closed"
      stop
  connected send = \case
    Msg msg@(TimestampedMessage t m) -> do
      case m of
        Ping -> univ ! (Msg (TimestampedMessage t (ClientId n)))
        Cmd _ cmd -> univ ! (Msg (TimestampedMessage t (Cmd n cmd))) 
        _ -> liftIO $ send msg
      become $ connected send
    Disconnect _ -> do
      univ ! Disconnect n
      log Info "connection closed"
      stop
    
